### PR TITLE
update ButtonIcon to TS

### LINKED ISSUE #18327: Migrate components to TS: ButtonIcon
### Description
Migrate `ButtonIcon` component to TypeScript

### Technical Details
- Convert Objects in `SNAKE_CASE` to enums in `PascalCase` follow [TS conventions](https://www.notion.so/Use-TypeScript-9ec7bd8fc8cf4d82b81cd8943c4707cd?pvs=4#7e6410b21100455a9a08442108f19d78)
- Update file name from `component-name.contants.js` to `component-name.types.ts` and add types
- Update file name from `component-name.js` to `component-name.tsx` and add types
- Update file name from `component-name.test.js` to `component-name.test.tsx`
- Update file name from `component-name.stories.js` to `component-name.stories.tsx`
- Update all instances of component objects to enums
- Update all documentation 

### Acceptance Criteria
- Component has been converted to TypeScript

### PR DESCRIPTION
**Pull request by garrettbear**
## Explanation

Update ButtonIcon to TS and move JS version to a deprecated folder
* Fixes #18327 

For any existing production components using ButtonIcon, the import was updated to point toward the deprecated folders. 
_Before_
`import { ButtonIcon } from '../../component-library';`
_After_
`import { ButtonIcon } from '../../component-library/button-icon/deprecated';`

## Screenshots/Screencaps

No visual UI change

## Manual Testing Steps
`yarn jest ui/components/component-library/button-icon`
`yarn jest ui/components/component-library`

## Pre-merge author checklist

- [X] I've clearly explained:
  - [X] What problem this PR is solving
  - [X] How this problem was solved
  - [X] How reviewers can test my changes
- [X] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [X] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [X] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

### CODE REVIEW

File: ui/components/component-library/button-icon/deprecated/button-icon.test.js [hunk 1]
```diff
@@ -1,8 +1,8 @@
 /* eslint-disable jest/require-top-level-describe */
 import { render } from '@testing-library/react';
 import React from 'react';
-import { IconColor } from '../../../helpers/constants/design-system';
-import { IconName } from '..';
+import { IconColor } from '../../../../helpers/constants/design-system';
+import { ICON_NAMES } from '../../icon/deprecated';
 import { BUTTON_ICON_SIZES } from './button-icon.constants';
```
Comment (NidhiKJha) [inline ui/components/component-library/button-icon/deprecated/button-icon.test.js:6]: The changes look really good to me. Before I approve it, one quick question - Should we update BUTTON_ICON_SIZES as well since this is a part of ButtonIcon??

**Review by NidhiKJha — DISMISSED** [review]

File: ui/components/component-library/button-icon/button-icon.tsx [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        className={classnames(
+          'mm-button-icon',
+          `mm-button-icon--size-${String(size)}`,
+          {
+            'mm-button-icon--disabled': Boolean(disabled),
+          },
+          className,
+        )}
+        color={color}
+        {...(disabled ? { disabled: true } : {})}
```
[the remark this answers is not in the document]
Comment (georgewrmarshall) [reply ui/components/component-library/button-icon/button-icon.tsx:48]: Shouldn't we be able to pass disabled as a prop whether it is `true` or `false`?

File: ui/components/component-library/button-icon/button-icon.tsx [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        display={DISPLAY.INLINE_FLEX}
+        justifyContent={JustifyContent.center}
+        alignItems={AlignItems.center}
+        borderRadius={BorderRadius.LG}
+        backgroundColor={BackgroundColor.transparent}
+        {...(href ? { href } : {})}
```
[the remark this answers is not in the document]
Comment (georgewrmarshall) [reply ui/components/component-library/button-icon/button-icon.tsx:54]: Also feels like this should be resolved through a type some how. Is this a pattern that exists in other libraries?

File: ui/components/component-library/button-icon/index.ts [hunk 4]
```diff
@@ -0,0 +1,2 @@
+export { ButtonIcon } from './button-icon';
+export { ButtonIconSize } from './button-icon.types';
```
Comment (georgewrmarshall) [inline ui/components/component-library/button-icon/index.ts:2]: We can export ButtonIconProps from here just not the root index.js file

**Review by georgewrmarshall — COMMENTED** [review]
Looking good! Left some comments will rereview once I touch down in Van

File: ui/components/component-library/text-field/README.mdx [hunk 5]
```diff
@@ -90,8 +90,9 @@ Use the `startAccessory` and `endAccessory` props to add components such as icon
 
 ```jsx
 import { Color, IconColor, SIZES, DISPLAY } from '../../../helpers/constants/design-system';
-import { Icon, IconName } from '../../component-library/deprecated'
-import { ButtonIcon, TextField } from '../../component-library';
+import { Icon, ICON_NAMES } from '../../component-library/icon/deprecated'
+import { ButtonIcon } from '../../component-library/button-icon/deprecated';
+import { TextField } from '../../component-library';
```
Comment (georgewrmarshall) [inline ui/components/component-library/text-field/README.mdx:95]: I made a mistake here in the clean up PR
```suggestion
import { ButtonIcon } from '../../component-library/button-icon/deprecated';
import { TextField, Icon, IconName } from '../../component-library';
```

File: ui/components/ui/box/box.d.ts [hunk 6]
```diff
@@ -131,6 +131,14 @@ export type BackgroundColorArray = [
   BackgroundColor?,
 ];
 
+interface BoxDisabledProp {
+  disabled?: boolean;
+}
+
+interface BoxHrefProp {
+  href?: string;
+}
```
[the remark this answers is not in the document]
Comment (georgewrmarshall) [reply ui/components/ui/box/box.d.ts:140]: Hmm Box should accept all html element attributes. If we want to merge this we should create an issue to fix the main issue otherwise we will be creating interface props for every html attribute possible?

File: ui/components/ui/box/box.d.ts [hunk 7]
```diff
@@ -327,7 +335,12 @@ export interface BoxProps extends React.HTMLAttributes<HTMLElement> {
   /**
    * Use the as prop to change the underlying HTML element of the Box component.
    */
-  as?: keyof HTMLElementTagNameMap;
+  as?:
+    | ('button' & BoxDisabledProp)
+    | ('input' & BoxDisabledProp)
+    | ('select' & BoxDisabledProp)
+    | ('a' & BoxHrefProp)
+    | keyof JSX.IntrinsicElements;
```
Comment (georgewrmarshall) [inline ui/components/ui/box/box.d.ts:343]: This doesn't seem right we shouldn't need to isolate every html element

File: ui/components/component-library/button-icon/button-icon.types.ts [hunk 8]
```diff
@@ -0,0 +1,49 @@
+import type { BoxProps } from '../../ui/box/box.d';
+import { IconName, IconSize } from '../icon';
+import { Color, IconColor } from '../../../helpers/constants/design-system';
+
+export enum ButtonIconSize {
+  Sm = IconSize.Sm,
+  Lg = IconSize.Lg,
+}
```
Comment (georgewrmarshall) [inline ui/components/component-library/button-icon/button-icon.types.ts:8]: Let's just use static strings until we are able to find a way to map or reuse enums
```suggestion
export enum ButtonIconSize {
  Sm = 'sm',
  Lg = 'lg',
}
```

File: ui/components/component-library/button-icon/button-icon.types.ts [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+export interface ButtonIconProps extends BoxProps {
+  /**
+   * String that adds an accessible name for ButtonIcon
+   */
+  ariaLabel: string;
+  /**
+   * The polymorphic `as` prop allows you to change the root HTML element of the Button component between `button` and `a` tag
+   */
+  as?: keyof JSX.IntrinsicElements;
```
Comment (georgewrmarshall) [inline ui/components/component-library/button-icon/button-icon.types.ts:18]: Maybe if we are more strict on what the root html element can be we can have better control over what html attributes can be passed to the `ButtonIcon` component. Let's prevent changing the root element from anything but a `button` | `as`? I can't think of any reason that you would need to change it to anything else?

```suggestion
  as?: 'button' | 'a';
```

File: ui/components/component-library/button-icon/button-icon.types.ts [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+  /**
+   * An additional className to apply to the ButtonIcon.
+   */
+  className?: string;
+  /**
+   * The color of the ButtonIcon component should use the IconColor object from
+   * ./ui/helpers/constants/design-system.js
+   */
+  color?: IconColor | Color;
```
Comment (georgewrmarshall) [inline ui/components/component-library/button-icon/button-icon.types.ts:27]: Let's only allow `IconColor` we don't allow any custom icons or to allow for the icon to be a letter right?
```suggestion
  color?: IconColor;
```

File: ui/components/component-library/button-icon/button-icon.types.ts [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+   */
+  href?: string;
+  /**
+   * The name of the icon to display. Should be one of IconName
+   */
+  iconName: IconName;
+  /**
+   * iconProps accepts all the props from Icon
+   */
+  iconProps?: object;
```
Comment (georgewrmarshall) [inline ui/components/component-library/button-icon/button-icon.types.ts:43]: Can this be changed to `IconProps` now?
`import { IconProps } from '../icon';`
```suggestion
  iconProps?: IconProps;
```

File: ui/components/component-library/button-icon/button-icon.tsx [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+        className={classnames(
+          'mm-button-icon',
+          `mm-button-icon--size-${String(size)}`,
+          {
+            'mm-button-icon--disabled': Boolean(disabled),
+          },
+          className,
+        )}
+        color={color}
+        {...(disabled ? { disabled } : {})}
```
Comment (georgewrmarshall) [inline ui/components/component-library/button-icon/button-icon.tsx:48]: No that we are limiting `as` to just `button` and `a` tags could we do something like
`const isDisabled = disabled && Tag === 'button';`
```suggestion
        ...(isDisabled ? { disabled: true } : {})} // only allow disabled attribute to be passed down to the Box when the as prop is equal to a button element
```

File: ui/components/component-library/button-icon/button-icon.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (georgewrmarshall) [reply ui/components/component-library/button-icon/button-icon.tsx:54]: Ah I see. So we shouldn't pass a disabled attribute unless we can be sure that the root html element can accept it. I think this is fine. Looks like it's an issue with other libraries that solution is outside the scope of this PR. That's gonna be a whopper haha 

This is how Chakra handle it. They have their own forwardRef function which has a bit more to it to account for polymorphic props 
https://chakra-ui.com/community/recipes/as-prop

File: ui/components/component-library/button-icon/button-icon.tsx [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+        display={DISPLAY.INLINE_FLEX}
+        justifyContent={JustifyContent.center}
+        alignItems={AlignItems.center}
+        borderRadius={BorderRadius.LG}
+        backgroundColor={BackgroundColor.transparent}
+        {...(href ? { href } : {})}
+        ref={ref}
+        {...props}
+      >
+        <Icon name={iconName} size={size} {...iconProps} />
```
Comment (georgewrmarshall) [inline ui/components/component-library/button-icon/button-icon.tsx:58]: Looked this up and we could do something like this
```
const buttonIconSizeToIconSize: Record<ButtonIconSize, IconSize> = {
  [ButtonIconSize.Sm]: IconSize.Sm,
  [ButtonIconSize.Lg]: IconSize.Lg,
};

export const ButtonIcon = React.forwardRef(
  (
   ...
```
```suggestion
size={buttonIconSizeToIconSize[size]}
```

File: ui/components/component-library/icon/icon.tsx [hunk 14]
```diff
@@ -16,7 +16,11 @@ export const Icon = ({
   ...props
 }: IconProps) => (
   <Box
-    className={classnames(className, 'mm-icon', `mm-icon--size-${size}`)}
+    className={classnames(
+      className,
+      'mm-icon',
+      `mm-icon--size-${String(size)}`,
```
[the remark this answers is not in the document]
Comment (georgewrmarshall) [reply ui/components/component-library/icon/icon.tsx:22]: Shouldn't need this after you update the ButtonIconSize enum to use plain strings

File: ui/components/component-library/icon/icon.types.ts [hunk 15]
```diff
@@ -160,7 +160,7 @@ export interface IconProps extends BoxProps {
    * Possible values could be IconSize.Xs (12px), IconSize.Sm (16px), IconSize.Md (20px), IconSize.Lg (24px), IconSize.Xl (32px), IconSize.Inherit (inherits font-size).
    * Default value is IconSize.Md (20px).
    */
-  size?: IconSize;
+  size?: IconSize | keyof IconSize;
```
[the remark this answers is not in the document]
Comment (georgewrmarshall) [reply ui/components/component-library/icon/icon.types.ts:163]: Can remove this if that mapping const is used

**Review by georgewrmarshall — COMMENTED** [review]
I see what you mean haha. This is hard. I think our end game solutions for polymorphic props and typescript are outside the scope of this PR. Made some more suggestions and [created a PR with all of them](https://github.com/MetaMask/metamask-extension/pull/18503) that seems to be working

File: ui/components/ui/box/box.d.ts [hunk 16]
```diff
@@ -327,7 +326,7 @@ export interface BoxProps extends React.HTMLAttributes<HTMLElement> {
   /**
    * Use the as prop to change the underlying HTML element of the Box component.
    */
-  as?: keyof HTMLElementTagNameMap;
+  as?: keyof JSX.IntrinsicElements;
```
Comment (georgewrmarshall) [inline ui/components/ui/box/box.d.ts:329]: Why is this needed?

From ChatGPT: In typescript what is the difference between HTMLElementTagNameMap and JSX.IntrinsicElements?
Both HTMLElementTagNameMap and JSX.IntrinsicElements are TypeScript interfaces that represent a mapping between HTML tag names and their respective element types.

The difference between them is that HTMLElementTagNameMap is a built-in TypeScript interface that maps all valid HTML tag names to their corresponding element types. It is mainly used when working with the DOM API directly in TypeScript.

On the other hand, JSX.IntrinsicElements is a TypeScript interface that is used when writing JSX in TypeScript. It maps all valid HTML tag names to their corresponding element types, but also includes additional JSX-specific attributes, such as className and style.

In other words, HTMLElementTagNameMap is used when working with the DOM API directly, while JSX.IntrinsicElements is used when writing JSX in TypeScript.

**Review by georgewrmarshall — COMMENTED** [review]
Looking good. Left one question.

**Review by georgewrmarshall — APPROVED** [review]

**Review by NidhiKJha — APPROVED** [review]
